### PR TITLE
4.2/pagination

### PR DESCRIPTION
**Pull request by weeryan17**
- [ ] Completed PR - Has all features or fixes that it was meant to and isn't half done
    - The Utils are mostly done, but not everything has moved yet
- [x] Follows the style guide - Style guide found [here](https://github.com/FlareBot/FlareBot/blob/master/CONTRIBUTING.md)
- [x] All changes have been tested

## Added
PaginationUtil
PaginationList
The pagination system in general

## Changed
Moved Reports over to the new list system
Added message to buttons so we can do editing

### CODE REVIEW

File: src/main/java/stream/flarebot/flarebot/commands/moderation/RolesCommand.java [hunk 1]
```diff
@@ -58,7 +59,15 @@ public void onCommand(User sender, GuildWrapper guild, TextChannel channel, Mess
...
-            MessageUtils.sendInfoMessage(sb.toString(), channel, sender);
+            MessageUtils.sendInfoMessage(sb.toString(), channel, sender);*/
+
+            StringBuilder sb = new StringBuilder();
+            for(Role r : guild.getGuild().getRoles())
+                sb.append(r.getName()).append(" (").append(r.getId()).append(")\n");
+            int page = 1;
+            if(args.length == 1)
+                page = Integer.parseInt(args[0]);
+            //PaginationUtil.sendPagedMessage(channel, sb.toString(), PaginationUtil.SplitMethod.CHAR_COUNT, 1000, page, true);
```
Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/commands/moderation/RolesCommand.java]: Why is the commented?

File: src/main/java/stream/flarebot/flarebot/commands/moderation/RolesCommand.java [hunk 2]
```diff
@@ -19,7 +20,7 @@
     @Override
     public void onCommand(User sender, GuildWrapper guild, TextChannel channel, Message message, String[] args, Member member) {
         if (args.length <= 1) {
-            StringBuilder sb = new StringBuilder();
+            /*StringBuilder sb = new StringBuilder();
```
Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/commands/moderation/RolesCommand.java]: Don't leave large blocks of commented code

File: src/main/java/stream/flarebot/flarebot/commands/moderation/RolesCommand.java (hunk 1 above)
Comment (weeryan17) [reply src/main/java/stream/flarebot/flarebot/commands/moderation/RolesCommand.java]: From testing. give me a min to clean that up

File: src/main/java/stream/flarebot/flarebot/commands/moderation/ReportsCommand.java [hunk 3]
```diff
@@ -34,35 +35,29 @@ public void onCommand(User sender, GuildWrapper guild, TextChannel channel, Mess
...
+                        ArrayList<String> header = new ArrayList<>();
+                        header.add("Id");
+                        header.add("Reported");
+                        header.add("Time");
+                        header.add("Status");
                         List<Report> reports = guild.getReportManager().getReports();
-                        int page = 1;
-                        final int reportsLength = 15;
+
+                        if (reports.isEmpty()) {
```
Comment (WalshyDev) [inline src/main/java/stream/flarebot/flarebot/commands/moderation/ReportsCommand.java]: Do checks before putting a list into memory

File: src/main/java/stream/flarebot/flarebot/commands/moderation/ReportsCommand.java [hunk 4]
```diff
@@ -34,35 +35,29 @@ public void onCommand(User sender, GuildWrapper guild, TextChannel channel, Mess
...
+                            MessageUtils.sendInfoMessage("No Reports for this guild!", channel, sender);
+                            return;
+                        }
+
+                        List<List<String>> body = getReportsTable(reports);
+                        int page = 0;
                         if (args.length == 2) {
                             try {
                                 page = Integer.valueOf(args[1]);
```
Comment (WalshyDev) [inline src/main/java/stream/flarebot/flarebot/commands/moderation/ReportsCommand.java]: Replace with use of GeneralUtils#getInt

File: src/main/java/stream/flarebot/flarebot/util/PaginationList.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+        /**
+         * Gets the group as a string
+         * Useful for if you want to use the groups as pages
+         *
+         * @return The group as a string.
+         */
+        public String getGroupAsString() {
+            StringBuilder builder = new StringBuilder();
+            for (T page: group) {
```
Comment (WalshyDev) [inline src/main/java/stream/flarebot/flarebot/util/PaginationList.java]: Missing space

File: src/main/java/stream/flarebot/flarebot/util/PaginationUtil.java [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+            bodyBuilder.append(String.format(formatLine.toString(), row.toArray()));
+        }
+        String bodyString = bodyBuilder.toString();
+        List<String> pages = new ArrayList<>();
+        String[] lines = bodyString.split("\n");
+        int pagesCount = Math.max((int) Math.ceil((double) lines.length / rows), 1);
+        for (int i = 0; i < pagesCount; i++) {
+            String[] page = ArrayUtils.subarray(lines, rows * i, (rows * i) + rows);
+            StringBuilder sb = new StringBuilder();
+            sb.append("```");
```
Comment (WalshyDev) [inline src/main/java/stream/flarebot/flarebot/util/PaginationUtil.java]: Always make sure to \n after the start!

File: src/main/java/stream/flarebot/flarebot/util/PaginationUtil.java [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+            sb.append(MessageUtils.appendSeparatorLine("+", "+", "+", padding, widths));
+            sb.append(headerString);
+            sb.append(MessageUtils.appendSeparatorLine("+", "+", "+", padding, widths));
+            for (String line : page) {
+                sb.append(line).append("\n");
+            }
+            sb.append(MessageUtils.appendSeparatorLine("+", "+", "+", padding, widths));
+            sb.append(MessageUtils.getFooter("Page: " + (i + 1) + "/" + pagesCount, padding, widths));
+            sb.append(MessageUtils.appendSeparatorLine("+", "+", "+", padding, widths));
+            sb.append("```");
```
Comment (WalshyDev) [inline src/main/java/stream/flarebot/flarebot/util/PaginationUtil.java]: \n before the end.

File: src/main/java/stream/flarebot/flarebot/commands/moderation/ReportsCommand.java [hunk 8]
```diff
@@ -40,22 +40,18 @@ public void onCommand(User sender, GuildWrapper guild, TextChannel channel, Mess
                         header.add("Reported");
                         header.add("Time");
                         header.add("Status");
```
Comment (WalshyDev) [inline src/main/java/stream/flarebot/flarebot/commands/moderation/ReportsCommand.java]: I meant this list but below is also true

**Pull request merged**